README.md
PLM is a Node.js library for bidirectional, event-rich communication with INSTEON PowerLinc USB and RS232 modems.  
This documentation is not yet complete, and some parts of the API are subject to change without notice. Check CHANGELOG.md before updating.

Quick Start
===========
This example is a heavily-simplified version of the PLMsh script whose source can be found at `test.js`.

	var PLM = require("plm");
	
	var modem = new PLM("/dev/ttyUSB0", function(data){
		// Called when the modem is ready;
		// data contains the modem's device info
		console.log(
			"Connected to PLM\n" +
			"ID: " + PLM.util.makeHex(data.id) + "\n" +
			"Device Category: " + PLM.util.makeHex(data.deviceCategory) + "\n" +
			"Device Subcategory: " + PLM.util.makeHex(data.deviceSubcategory) + "\n" +
			"Firmware Version: " + PLM.util.makeHex(data.firmwareVersion)
		);
		// Send an INSTEON command
		modem.sendINSTEON({
			to: PLM.util.parseHex("AB.CD.EF"), // Device ID, separator optional
			command: util.parseHex("11FF", 2) // INSTEON command (Light on at full)
		}, function(err, reply){
			if(err){
				console.error(err); // Something's wrong!
			}else{
				console.log("REPLY FROM: " + PLM.util.makeHex(reply.from)); // "AB.CD.EF"
				console.log("COMMAND NUMBER: " + PLM.util.makeHex(reply.command.number)); // "0x11"
				console.log("COMMAND DATA: " + PLM.util.makeHex(reply.command.data)); // "0xFF"
			}
		});
	});


Terms
=====
Below are some brief explanations of some of the terms used in this document.

"PLM"
-----
The term "PLM" usually refers to a USB or RS232 [SmartLabs PowerLinc Modem](http://www.smarthome.com/2413S/PowerLinc-Modem-INSTEON-Serial-Interface-Dual-Band/p.aspx "Link to SmartHome page"); however, to avoid confusion with this module's name, the term "PLM" is not used to refer to the modem in this document. Here, `PLM` will refer to this module itself and its main entry-point class.

"IM"/"Modem"
------------
"IM" and "Modem" in body text refer to an INSTEON modem this module is compatible with. `modem` in a code span refers to an instance of this module.

"Device"/"Responder"
--------------------
"Device" and "Responder" refer to an INSTEON or X10 device being controlled via an IM using this module. The IM itself will be referred to by the terms above.

"Host"
------
"Host" refers to the computer connected to an IM using either RS232 or USB.

Compatability
=============
This module should be compatible with all SmartLabs PowerLinc Modems. It may or may not be compatible with SmartLabs PowerLinc Controllers (pending testing).

General tips
============
In the `PLM` API, many methods support multiple input types to make it easier to use the library with various different coding styles. There are also a few things that you need to keep in mind when working with binary data. Here are some general rules that should apply across the entire library:

`Buffer`s can be `Array`s
-------------------------
Any time you can pass a `Buffer` into a function, you can pass an `Array` of bytes as `Number`s instead, as `PLM` will call `new Buffer(argument)`.

`INSTEONMessage`s can be `Array`s, `Buffer`s, or compatible `Object`
--------------------------------------------------------------------
Any time you can pass an `INSTEONMessage` into a function, you can pass a `Buffer`, `Array` of bytes, or `Object` with compatible keys, as `PLM` will call `new INSTEONMessage(argument)`.

`INSTEONMessageFlags` can be a `Number`, 1-byte `Buffer` or `Array`, or compatible `Object`
-------------------------------------------------------------------------------------------
Any time you can pass an `INSTEONMessageFlags` into a function, you can pass a `Number` representing a bitmask, a 1-byte `Array` or `Buffer`, or an `Object` with compatible keys, as `PLM` will call `new INSTEONMessageFlags(argument)`.

INSTEON `command`s can be `Buffer`s or `Object`s or `Number`s
-------------------------------------------------------------
In `modem.sendINSTEON(message, callback)`, `message.command` can be any of 3 types:
- A `Buffer` containing the command 1 and command 2 fields. A 1-byte `Buffer` will result in `0x00` in command 2
- An `Object` in the format `{number: 0x11, data: 0xFF}`, `command.number` corresponding to the command 1 field and `command.data` corresponding to command 2 and defaulting to `0x00`
- A `Number` representing the Command 1 field; Command 2 will be `0x00`

Modem commands can leave off `STX (0x02)`
---------------------------------------
All actual raw serial commands to the modem start with `STX (0x02)`, but if you don't want to pass that in, `PLM` will add it for you. This applies to all direct-command functions (e.g. `modem.sendCommand`, `modem.sendCommandNow`), but not to `modem.sendRaw`.

Unless otherwise specified, too-short `Buffer`s may result in garbage data
--------------------------------------------------------------------------
Most functions do not check the lengths of the `Buffer`s they are passed, assuming that either your buffers are correct-length or you're OK with the consequences. If you pass in a `Buffer` with a length shorter than expected, `PLM` will not fill the remaining space with `0x00`, so the remainder of the field may be filled with garbage data that remains in memory. Only use too-short `Buffer`s if your command definitely doesn't parse data past a certain point; if you're unsure, use a correct-length `Buffer` and fill the remainder with `0x00` yourself.

API
===
This module contains multiple classes, each of which has a set of methods and properties. Below is an explanation of the usage of each.

Class: PLM
----------
The `PLM` class can be accessed using `require("plm")`. It's the root class for the entire library, and you connect to an INSTEON modem by instantiating it. These docs will assume that you call `var PLM = require("plm");`.

### Constructor
Calling `new PLM("/path/to/modem/devfs/entry", callback)` (NOTE: `PLM` will NOT attempt to auto-detect your modem's serialport. This may change in the future, but for now, assume it won't) will connect `PLM` to the modem using the `node-serialport` module. These docs will refer to a `PLM` instance as `modem`.  
Once it's connected, `PLM` will send your modem the command `GET_IM_INFO (0x60)`. The modem will reply with some information about itself, including its hardcoded INSTEON Device ID, its Device Category, its Device Subcategory, and its firmware version. An object containing these will be passed as the first argument to the `callback` function; we'll assume you called the arg `data`.  
`data.id` is a 3-byte `Buffer`. `data.deviceCategory`, `data.deviceSubcategory`, and `data.firmwareVersion` are `Number`s.  
Once the `callback` has been called, `PLM` is ready to send and receive INSTEON and X10 commands.

### `modem.busy`
This `Boolean` keeps track of whether or not the modem is currently busy with a command.  
DO NOT MODIFY THIS VALUE; this may result in unexpected behavior.  
You probably don't have to worry about this, as `modem.sendCommand` and `modem.dequeue` deal with it for you.

### `modem.queue`
This `Array` is the queue of commands to be sent to the modem.  
Each item is an `Object` with format `{command: Buffer([...command data...]), callback: function(error, replyData, commandNumber){}}`; the callback is optional, and will be called when the modem replies to the command or times out.  
You probably don't have to worry about this, as `modem.sendCommand` and `modem.dequeue` deal with it for you.

### `modem.INSTEONQueue`
This `Array` represents the INSTEON message queue. Elements are in the format `{message: INSTEONMessage({...message data...}), callback: function(error, message){}}`; the callback is optional, and will be called when the modem sends an error, the device replies, or the command times out.  
You probably don't have to worry about this, as `modem.sendINSTEON` and `modem.dequeue_INSTEON` deal with it for you.

### `modem.commandTimeout`
This `Number` represents the amount of time, in milliseconds, that `PLM` will wait for a response from a modem. `250` milliseconds is the default, as it's just above the amount of time after which the modem itself times out and clears its message buffer by default. If a response is not received after this time, your command's `callback`, if present, will be called with its first argument as `MODEM_TIMEOUT` `Error`. Note that receiving a timeout callback is not a perfect guarantee that the command failed, a timeout will result in an automatic dequeue, and if a command times out and later a reply is received, the callback will not fire again. Set this to `false` to disable modem timeouts (this may result in unexpected behavior).

### `modem.minCommandInterval`
This `Number` represents the minimum amount of time, in milliseconds, between the PLM acknowledging a previous command and the library advancing its queue and sending the next one. It defaults to `250`; this is rather arbitrary and is subject to change. A value of `0` disables this timer.

### `modem.INSTEONStandardTimeout`
This `Number` represents the amount of time, in milliseconds, that `PLM` will wait for a response to an INSTEON direct standard-length message. `2500` milliseconds is the default; this is rather arbitrary and subject to change. This timeout is only set if a `callback` is provided to `modem.sendINSTEON`. Note that receiving a timeout callback is not a perfect guarantee that the command failed, and if the device timeout occurs, and the reply is later received, the callback will not fire again. Set this to `false` to disable device timeouts (this may result in unexpected behavior).

### `modem.INSTEONExtendedTimeout`
This `Number` is identical to `modem.INSTEONStandardTimeout`, except that it's used for extended messages rather than standard-length ones. It defaults to `5000` milliseconds; this is rather arbitrary and subject to change.

### `modem.minINSTEONInterval`
This `Number` represents the minimum amount of time, in milliseconds, between receiving an `ACK`/`NAK` for an INSTEON command and the library advancing its queue and sending the next one. It defaults to `0`, which disables this timer.

### `modem.X10Timeout`
This `Number` is identical to `modem.INSTEONStandardTimeout`, except that it's used for X10 messages rather than INSTEON ones. Note that X10 is not currently implemented, so this value has no effect.

### `modem.sendINSTEON(message, callback)`
This method sends an INSTEON direct message from the modem. The message should be an `INSTEONMessage`. `callback` will be called when either the modem returns an error, the device replies, or either the modem or the device times out. The arguments are `callback(error, message)`, where `message` is the `INSTEONMessage` `ACK` sent by the device and `error` is either false (if the operation was successful) or one of `MODEM_NAK`, `MODEM_TIMEOUT`, `DEVICE_NAK`, or `DEVICE_TIMEOUT`. If the error is `DEVICE_NAK`, the `message` will still be provided. For other errors, `message` will be `undefined`.

### `modem.sendCommand(command, callback, jump)`
This method queues a command to be sent over the serial link to the modem, and tries to advance the queue. If the modem is busy with another command, this will return `false`, but the message will still be queued. If the `callback` arg is present, `callback(error, replyData, commandNumber)` will be called when the modem sends a reply. Arguments are:
- `error` is either `false` or an error (either `MODEM_NAK` or `MODEM_TIMEOUT`),
- `replyData` is a `Buffer` containing the modem's response data, not including the `STX`, command number, or `ACK/NAK` bytes.
- `commandNumber` is a `Number` which should match the first byte of your original `command`.
By default, `sendCommand` will `push` new commands on the end of the queue, but if your command is high-priority, you can set `jump` to `true`, and your command will be `unshift`ed onto the front of the queue.

### `modem.dequeue()`
This method attempts to send the first item in the command queue to the modem. If the modem is busy, this method will return `false`. If there are no items in the queue, it will return `undefined`. Otherwise, it will call `modem.sendCommandNow` with the first item in the queue, prepare event listeners and timeouts for the reply, remove the item, and return `true`. You shouldn't have to worry about calling this, as `PLM` calls it automatically.

### `modem.INSTEONDequeue()`
This method is identical to `modem.dequeue`, except that it advances the INSTEON queue, rather than the modem command queue.

### `modem.sendCommandNow(command, callback)`
This method prepends `STX (0x02)` to the `command` if necessary, then sends it directly to the modem using `modem.sendRaw`. The `callback` parameter will be passed to `modem.sendRaw`; it's only used to start timers internally, as it only fires when the command has been sent, not when a reply has been received.  
NOTE: YOU PROBABLY SHOULD NOT CALL THIS METHOD DIRECTLY. THE MODEM CANNOT HANDLE MULTIPLE CONCURRENT COMMANDS.  
The queue system keeps track of the modem's state and queues additional commands when the modem is busy. If a command is sent immediately without checking the modem's status, unexpected behavior may result. You have been warned.

### `modem.sendINSTEONNow(message, callback)`
This method generates a modem command `Buffer` from an `INSTEONMessage` and queues it to be sent to the modem without adding it to the INSTEON message queue. You most likely don't have to worry about this method, as `modem.INSTEONDequeue` calls it automatically. Incorrect use of this method will not cause modem errors, but may cause INSTEON messages to be dropped due to interference on the power line from other messages. `modem.sendINSTEON` is reccomended.

### `modem.sendRaw(data, callback)`
This method sends `data` directly to the modem without any preprocessing, queueing, or checking. `callback`, if present, will be called when the command has been sent (not when a reply has been received).  
NOTE: YOU PROBABLY SHOULD NOT CALL THIS METHOD DIRECTLY. THE MODEM CANNOT HANDLE MULTIPLE CONCURRENT COMMANDS. THE MODEM MAY NOT RESPOND CORRECTLY TO COMMANDS NOT BEGINNING WITH `STX (0x02)`.  
The queue system keeps track of the modem's state and queues additional commands when the modem is busy. If a command is sent immediately without checking the modem's status, unexpected behavior may result. You have been warned.

Class: `INSTEONMessage`
---------------------
`INSTEONMessage` represents a message sent or to be sent by an IM or other device over the INSTEON network. This class is available at `PLM.INSTEONMessage`, and inherits from `Buffer`.

### Constructor
Calling `new INSTEONMessage(data)` returns a new INSTEON message with the data specified. Data can be any of:
* A `Buffer` containing a pre-formatted INSTEON message for the IM, either for sending or receiving.
* An `Object` with compatible keys, including an existing `INSTEONMessage` instance.
This documentation will refer to an instance of `INSTEONMessage` as `message`.
An `INSTEONMessage`'s `Buffer` value can be used directly when sending a message to an IM, but will not match the content of a received message (the 3-byte "from" field is missing, deliberately). To compare two messages, use `message.isEqual`.

### Class method: `INSTEONMessage.isINSTEONMessage(message)`
This method returns a `Boolean` representing whether or not `message` is an `INSTEONMessage`.

### 

Class: `INSTEONMessageFlags`
----------------------------
`INSTEONMessageFlags` represents the flags byte of an `INSTEONMessage`. This class is available at `PLM.INSTEONMessageFlags`, and inherits from `Number`.

Class: `INSTEONMessageCommand`
------------------------------
`INSTEONMessageCommand` represents the command bytes of an `INSTEONMessage`. This class is available at `PLM.INSTEONMessageCommand`, and inherits from `Buffer`.

Class: util
-----------


Constants
---------
Various constants are available in `lib/constants.js`, which is exposed as `PLM.CONSTANTS`. Not every constant is individually documented, but most should be self-explanatory.

### `MIN_OUT_COMMAND` (`0x60`)
All modem command numbers `>=` this constant are sent from the host machine to the modem. It's most likely only useful internally.

### `ACK` (`0x02`)
The first byte in any message to or from an IM must be `STX (0x02)`.

### `ACK` (`0x06`)
The `ACK` byte is used to signify that a command was successful.

### `NAK` (`0x15`)
The `NAK` byte is sent in place of `ACK` when a command fails.

### `PLM_COMMANDS`
This object contains each IM command's command number that can be sent to or received from an IM. For instance, `PLM_COMMANDS.INSTEON_STD_RECV == 0x50`.

### `CMD_LEN`
This object contains the number of bytes each IM command can be expected to contain, by command number, with the exception of `0x62` (`SEND_INSTEON`), whose length varies depending on if the message being sent is standard- or extended-length. It's most likely only useful internally. For instance, `CMD_LEN[0x50] == 9`.

### `INSTEON_COMMANDS`
This object contains the standard set of INSTEON commands, sorted first by type (`DIRECT`, `GROUP`, or `BROADCAST`), then by length (`STD` or `EXT`). For instance, `INSTEON_COMMANDS.DIRECT.STD.LIGHT_ON == 0x11`. `INSTEON_COMMANDS.DIRECT.EXT` is currently empty due to a case of lazy; pull requests are appreciated. Any inaccuracies or addition requests should be submitted to the Github issues page for this module.

### `BUTTON_EVENTS`
This object contains the event code for each of the button events an IM could send via command `0x54` (`BUTTON_EVENT_REPORT`). For instance, `BUTTON_EVENTS.SET_HELD == 0x03`.

TODO
====
- Add timeouts with sensible defaults
- Add a metric ton of convenience functions
- Finish documentation (esp. of subclasses)


CHANGELOG.md
CHANGELOG
=========

This document will be updated whenever the API changes.

0.3.2
-----
- Added minimum times between PLM and INSTEON commands

0.3.0
-----
- Added INSTEON queueing (hat tip to IceKarma)
- Bugfixes
- Improved package.json

0.2.1
-----
- Bugfixes
- Callback for `sendCommand`'s first argument is now `error`

0.2.0
-----
- Massive restructuring of every class
- Added timeouts

0.1.1
-----
- Initial documented version


test.js
#! /usr/bin/env node
var PLM = require("./"),
	util = PLM.util,
	readline = require("readline");

var rl = readline.createInterface({
	input: process.stdin,
	output: process.stdout,
	completer: completer,
	terminal: true
});

var modem = new PLM(process.argv[2], function(data){
	console.log(
		"Connected to PLM\n" +
		"ID: " + util.makeHex(data.id) + "\n" +
		"Device Category: " + util.makeHex(data.deviceCategory) + "\n" +
		"Device Subcategory: " + util.makeHex(data.deviceSubcategory) + "\n" +
		"Firmware Version: " + util.makeHex(data.firmwareVersion)
	);
	rl.prompt();
});

modem.on("error", function(err){
	console.error(err.toString());
	rl.prompt();
});

function completer(line){
	return [[], ""];
}

var helpCmds = {
	sendCommand: "",
	sendDirect: ""
}

var commands = {
	sendCommand: function(){
		rl.pause();
		modem.sendCommand(util.parseHex(Array.prototype.slice.call(arguments)), function(ACK, cmd, data){
			console.log("ACK:" + ACK + "; CMD:" + util.makeHex(cmd) + "; DATA: " + util.makeHex(data));
			rl.resume();
			rl.prompt();
		});
	},
	sendDirect: function(to, cmd, hops){
		rl.pause();
		modem.sendINSTEON({
			to: util.parseHex(to),
			command: util.parseHex(cmd, 2),
			flags: {
				maxHops: parseInt(hops, 10)
			}
		}, function(err, reply){
			if(err){
				console.error(err);
			}else{
				console.log("REPLY FROM: " + util.makeHex(reply.from));
				console.log("COMMAND NUMBER: " + util.makeHex(reply.command.number));
				console.log("COMMAND DATA: " + util.makeHex(reply.command.data));
			}
			rl.resume();
			rl.prompt();
		});
	},
	help: function(command){
		if(command in helpCmds){
			console.log(helpCmds[command]);
			rl.prompt();
		}else{
			console.log("Available commands: " + Object.keys(helpCmds).sort().join(", "));
			rl.prompt();
		}
	}
};

rl.on("line", function(line){
	var args = line.split(" ");
	if(args[0] in commands){
		commands[args[0]].apply(this, args.slice(1));
	}else{
		console.error("Command not found: " + args[0] + ". Type `help` for a list.");
		rl.prompt();
	}
});


lib/util.js
var CONSTANTS = require("./constants.js");
var util = module.exports = {
	makeBitmask: function(arr){
		var nMask = 0, nFlag = 0, nLen = arr.length > 32 ? 32 : arr.length;
		for (nFlag; nFlag < nLen; nMask |= arr[nFlag] << nFlag++);
		return nMask;
	},
	parseBitmask: function(nMask){
		var rMask;
		if(Buffer.isBuffer(nMask) || Array.isArray(nMask)){
			rMask = 0;
			for(var i = 0; i < nMask.length; i++){
				rMask |= nMask[0] << (8 * (nMask.length - i - 1));
			}
		}else{
			rMask = nMask;
		}
		// nMask must be between -2147483648 and 2147483647
		if (rMask > 0x7fffffff || rMask < -0x80000000) {
			throw new TypeError("arrayFromMask - out of range");
		}
		for (var nShifted = rMask, aFromMask = []; nShifted; aFromMask.push(Boolean(nShifted & 1)), nShifted >>>= 1);
		return aFromMask;
	},
	parseHex: function(hex, length, fill){
		if(!Array.isArray(hex)){
			hex = hex.replace(/$\s+|\s+^/, "").replace(/$0x/, "").replace(/([0-9A-F]{2})(?=[0-9A-F])/gi, "$1 ").split(/[^0-9A-F]+/i);
			// Strips whitespace and 0x, adds spaces every 2 hex characters in strings of >2, and splits by non-hex.
		}
		var buf = new Buffer(length || hex.length);
		for(var i = 0; i < buf.length; i++){
			buf[i] = hex[i] ? parseInt(hex[i], 16) : (fill || 0);
		}
		return buf;
	},
	makeHex: function(buffer, separator, lowercase){
		if(typeof buffer == "number"){
			buffer = new Buffer([buffer]);
		}else{
			buffer = new Buffer(buffer);
		}
		if(separator === undefined){
			if(buffer.length == 1){
				separator = " 0x";
			}else{
				separator = ".";
			}
		}else if(!separator){
			separator = "";
		}
		var str = "";
		for(var i = 0; i < buffer.length; i++){
			var hex = buffer.toString("hex", i, i + 1);
			str += separator + (lowercase ? hex : hex.toUpperCase());
		}
		return str.slice(1); // Kill the first separator
	},
	stripUntilStarts: function(buf, starts){
		while(buf[0] != starts && buf.length){
			// Something's gone horribly wrong and a buffer doesn't
			// start with a byte it should. Throw out bytes until it does,
			// or until we run out of buffer.
			buf = buf.slice(1);
		}
		return buf;
	},
	getMessageLength: function(buf){
		// Checks to see if the buffer contains any entire IM messages
		// returns the first one if it does
		if(buf.length == 0){
			// Fresh out of data.
			return false;
		}
		if(buf[0] != CONSTANTS.STX){
			// This should never throw, as starting with STX is enforced by the parent.
			throw new Error("Message buffer does not start with STX; something is FUBAR!");
		}
		if(buf.length < 2){
			return false;
		}
		var length = length = CONSTANTS.CMD_LEN[buf[1]];
		
		if(length === undefined){
			if(buf[1] == CONSTANTS.PLM_COMMANDS.SEND_INSTEON){
				// This is an echo/reply to an INSTEON message.
				// Check the flags to see if it's extended, then save the length.
				if(buf.length < 5){
					return false;
				}
				// Not using INSTEONMessageFlags to avoid a cyclic dependency!
				length = util.parseBitmask(buf[5])[4] ? 21 : 7;
			}else{
				throw new Error("Received unrecognized command; something is FUBAR! BUF: " + buf.toString("hex"));
			}
		}
		
		if(buf.length >= length + 2){
			return length + 2;
		}
		
		// No commands in the buffer; move on.
		return false;
	},
	buffersEqual: function(a, b){
		if(!(Buffer.isBuffer(a) && Buffer.isBuffer(b) && a.length === b.length)){
			return false;
		}
		
		for(var i = 0; i < a.length; i++){
			if (a[i] !== b[i]){
				return false;
			}
		}
		
		return true;
	}
};


lib/INSTEONMessageCommand.js
var INSTEONMessageCommand = module.exports = function INSTEONMessageCommand(cmd){
	if(!(this instanceof INSTEONMessageCommand)){ 
		return new INSTEONMessageCommand(cmd);
	}
	Buffer.call(this, 2);
	if(typeof cmd == "number"){
		this[0] = cmd;
		this[1] = 0;
	}else if(Buffer.isBuffer(cmd) || Array.isArray(cmd)){
		this[0] = cmd[0] || 0;
		this[1] = cmd[1] || 0;
	}else if(typeof cmd == "object"){
		this[0] = cmd.number || 0;
		this[1] = cmd.data || 0;
	}else{
		this[0] = this[1] = 0;
	}
};

INSTEONMessageCommand.prototype = new Buffer(0);
Object.defineProperties(INSTEONMessageCommand.prototype, {
	number: {
		get: function(){
			return this[0];
		},
		set: function(value){
			return this[0] = value;
		}
	},
	data: {
		get: function(){
			return this[1];
		},
		set: function(value){
			return this[1] = value;
		}
	}
});


lib/PLM.js
var events = require("events"),
//	stream = require("stream"), // Not using streams now; may later
	SerialPort = require("serialport").SerialPort,
	CONSTANTS = require("./constants.js"),
	util = require("./util.js"),
	INSTEONMessage = require("./INSTEONMessage.js"),
	INSTEONMessageFlags = require("./INSTEONMessageFlags.js"),
	INSTEONMessageCommand = require("./INSTEONMessageCommand.js");

var PLM = module.exports = function PLM(path, cb){
	if(!(this instanceof PLM)){ 
		return new PLM(path, cb);
	}
	var self = this;
	
	if(cb){
		this.once("connected", cb);
	}
	
	this.queue = [];
	
	this.INSTEONQueue = [];
	
	this.busy = this.INSTEONBusy = false;
	
	var serial = this.serial = new SerialPort(path, {
		// Serial port settings for standard IM (serial or USB)
		baudrate: 19200,
		databits: 8,
		parity: "none",
		stopbits: 1
	});
	
	serial.on("error", function(err){
		// Emit serialport errors on ourself
		self.emit("error", err);
	});
	serial.on("open", function(){
		// Send GET_IM_INFO to fetch our INSTEON ID
		self.sendCommand([CONSTANTS.PLM_COMMANDS.GET_IM_INFO], function(error, ret){
			if(error){
				return self.emit("error", error);
			}
			self.PLM_ID = ret.slice(0,3);
			self.emit("connected", {
				id: self.PLM_ID,
				deviceCategory: ret[3],
				deviceSubcategory: ret[4],
				firmwareVersion: ret[5]
			});
		});
	});
	
	serial.on("data", function(buf){
		self.emit("raw", buf);
	});
	
	this.rawBuffer = new Buffer(0);
	
	this.on("raw", function(buf){
		self.rawBuffer = Buffer.concat([this.rawBuffer, buf], this.rawBuffer.length + buf.length);
		var messageLength = 0;
		while(messageLength = util.getMessageLength(
			self.rawBuffer = util.stripUntilStarts(self.rawBuffer, CONSTANTS.STX)
		)){ // Search for all messages
			self.emit("message", self.rawBuffer.slice(0, messageLength + 1));
			self.rawBuffer = self.rawBuffer.slice(messageLength + 1); // Remove complete messages from buffer
		}
	});
	
	this.on("message", function(buf){
		if(buf[0] != CONSTANTS.STX){
			return false; // All messages should start with an STX byte
		}
		if(buf[1] < CONSTANTS.MIN_OUT_COMMAND){
			// Sent by IM; not an echo
			// Emit an event named by the command number
			self.emit(buf[1], buf.slice(2));
			self.emit("incoming", buf.slice(1));
		}else{
			// Echo/Reply to a command from host
			// Move ahead in queue
			var ACK = buf[buf.length - 1] == CONSTANTS.ACK;
			// Emit an echo event with the ACK/NAK status,
			// the command number, and the actual data
			self.emit("reply", ACK ? false : new Error("MODEM_NAK"), buf.slice(2, buf.length - 1), buf[1]);
			if(self.minCommandInterval){
				setTimeout(function(){
					self.busy = false;
					self.dequeue();
				}, self.minCommandInterval);
			}else{
				self.busy = false;
				self.dequeue();
			}
		}
	});
	
	function handleMsgEvent(buf){
		var msg = new INSTEONMessage(buf);
		// All messages get the generic event
		self.emit("INSTEON message", msg);
		if(!util.buffersEqual(msg.to, self.PLM_ID)){
			self.emit("INSTEON monitor message");
		}
		if(msg.flags.ACK){
			// Emit a generic ACK event with the whole message
			self.emit("INSTEON ACK", msg);
			if(msg.flags.group){
				self.emit("INSTEON group ACK", msg);
			}else{
				if(!util.buffersEqual(msg.to, self.PLM_ID)){
					self.emit("INSTEON monitor direct ACK");
					return; // Regular ACK: events don't fire for monitor mode
				}
				self.emit("INSTEON direct ACK", msg);
				// And one named for the device it came from
				// (Used for the PLM.sendINSTEON callback)
				self.emit("ACK:" + msg.from.toString("hex"), msg);
			}
		}else if(msg.flags.broadcast){
			self.emit("INSTEON broadcast message", msg);
		}else if(msg.flags.group){
			self.emit("INSTEON group message", msg);
		}else{
			if(!util.buffersEqual(msg.to, self.PLM_ID)){
				self.emit("INSTEON monitor direct message");
				return; // Regular message events don't fire for monitor mode
			}
			self.emit("INSTEON direct message", msg);
		}
	}
	
	this.on(CONSTANTS.PLM_COMMANDS.INSTEON_STD_RECV, handleMsgEvent);
	this.on(CONSTANTS.PLM_COMMANDS.INSTEON_EXT_RECV, handleMsgEvent);
	
	this.on(CONSTANTS.PLM_COMMANDS.ALL_LINKING_COMPLETED, function(data){
		this.emit("all_linking_completed", false, {
			type: data[0],
			group: data[1],
			id: data.slice(2, 5),
			category: data[5],
			subcategory: data[6],
			firmware_version: data[7]
		});
	});
	
	return this;
}

PLM.CONSTANTS = CONSTANTS;
PLM.util = util;
PLM.INSTEONMessage = INSTEONMessage;
PLM.INSTEONMessageFlags = INSTEONMessageFlags;
PLM.INSTEONMessageCommand = INSTEONMessageCommand;

PLM.prototype = new events.EventEmitter();

PLM.prototype.linkINSTEONDevice = function linkINSTEONDevice(mode, group, cb){
	var self = this;
	this.sendCommand([CONSTANTS.PLM_COMMANDS.START_ALL_LINKING, mode, group], function(err, reply_data, cmd_number){
		if(err){
			return cb(err);
		}
		self.once("all_link_finished", cb);
	});
};

PLM.prototype.cancelLinking = function cancelLinking(cb){
	var self = this;
	this.sendCommand([CONSTANTS.PLM_COMMANDS.CANCEL_ALL_LINKING], function(err, reply_data, cmd_number){
		if(err){
			return cb && cb(err);
		}
		self.emit("all_link_finished", (new Error("CANCELED")));
		cb(false);
	});
}

PLM.prototype.sendINSTEON = function sendINSTEON(message, cb, jump){
	var self = this;
	message = new INSTEONMessage(message);
	if(jump){
		this.INSTEONQueue.unshift({message: message, callback: cb});
	}else{
		this.INSTEONQueue.push({message: message, callback: cb});
	}
	return this.INSTEONDequeue();
};

PLM.prototype.sendINSTEONNow = function sendINSTEONNow(message, cb){
	var self = this;
	// Callback gets 2 args:
	// 1. Error
	// 2. Reply message object
	message = new INSTEONMessage(message);
	var data = new Buffer(message.length + 1);
	data[0] = CONSTANTS.PLM_COMMANDS.SEND_INSTEON; // Prefix the SEND_INSTEON command number
	message.copy(data, 1);
	this.sendCommand(data, cb);
};

PLM.prototype.sendCommand = function sendCommand(cmd, cb, jump){
	var buf = new Buffer(cmd);
	if(jump){
		this.queue.unshift({command: buf, callback: cb});
	}else{
		this.queue.push({command: buf, callback: cb});
	}
	return this.dequeue();
};

PLM.prototype.INSTEONDequeue = function INSTEONDequeue(){
	var self = this;
	if(this.INSTEONBusy){
		return false;
	}
	var cmd = this.INSTEONQueue.shift();
	if(!cmd){
		return;
	}
	var cb = cmd.callback;
	var message = cmd.message;
	var to = (new Buffer(message.to)).toString("hex");
	var extended = message.flags.extended;
	var timeoutDuration = (extended ? self.INSTEONExtendedTimeout : self.INSTEONStandardTimeout) || false;
	this.sendINSTEONNow(cmd.message, function(err, data, cmdnum){
		if(err){
			self.INSTEONBusy = false;
			self.INSTEONDequeue();
			if(cb){
				cb(err, data, cmdnum);
			}
			return;
		}
		// If everything's fine, set up a timeout (if enabled) and event listener
		var timeout, callback = function(reply){
			if(cb){
				if(reply.flags.NAK){
					cb(new Error("DEVICE_NAK"), reply);
				}else{
					cb(false, reply);
				}
			}
			if(timeoutDuration){
				clearTimeout(timeout);
			}
			self.removeListener("ACK:" + to, callback);
			if(self.minINSTEONInterval){
				setTimeout(function(){
					self.INSTEONBusy = false;
					self.INSTEONDequeue();
				}, self.minINSTEONInterval);
			}else{
				self.INSTEONBusy = false;
				self.INSTEONDequeue();
			}
		};
		self.on("ACK:" + to, callback);
		if(timeoutDuration){
			timeout = setTimeout(function(){
				self.INSTEONBusy = false;
				self.removeListener("ACK:" + to, callback);
				if(cb){
					cb(new Error("DEVICE_TIMEOUT"));
				}
				self.INSTEONDequeue();
			}, timeoutDuration);
		}
	});
	return this.INSTEONBusy = true;
}

PLM.prototype.dequeue = function dequeue(){
	var self = this;
	if(this.busy){
		return false; // Refuse to dequeue when the PLM is busy
	}
	var cmd = this.queue.shift();
	if(!cmd){
		return; // Queue was empty
	}
	this.sendCommandNow(cmd.command, function(){
		// Deal with timeouts if we have to
		if(self.commandTimeout){
			var timeout, callback = function(err, replyData, commandNumber){
				// "reply" event listener
				if(cmd.callback){
					// If we've got an actual callback function, call it
					cmd.callback(err, replyData, commandNumber);
				}
				// Event came. Clear timeout and remove this listener.
				clearTimeout(timeout);
				self.removeListener("reply", callback);
			}
			timeout = setTimeout(function(){
				// Event didn't come in time.
				if(cmd.callback){
					// If we've got an actual callback function, call it
					cmd.callback(new Error("MODEM_TIMEOUT"), false, false);
				}
				// Remove the regular callback so we don't double up.
				self.removeListener("reply", callback);
				// Advance queue; there's a chance this'll cause more errors 0.o
				self.dequeue();
			}, self.commandTimeout);
			self.on("reply", callback);
		}else{
			// If timeouts are disabled, just use `.once`
			self.once("reply", cmd.callback);
		}
	});
	return this.busy = true;
};

PLM.prototype.commandTimeout = PLM.prototype.minCommandInterval = 250;

PLM.prototype.minINSTEONInterval = 0; // This shouldn't be necessary, but provide the feature just in case

PLM.prototype.INSTEONStandardTimeout = PLM.prototype.X10Timeout = 2500;

PLM.prototype.INSTEONExtendedTimeout = 5000;

PLM.prototype.sendCommandNow = function sendCommandNow(cmd, cb){
	var data;
	if(cmd[0] !== CONSTANTS.STX){
		data = new Buffer(cmd.length + 1);
		data[0] = CONSTANTS.STX;
		cmd.copy(data, 1);
	}else{
		data = cmd;
	}
	this.sendRaw(data, cb);
}

PLM.prototype.sendRaw = function sendRaw(data, cb){
	this.serial.write(new Buffer(data), cb);
}
